dev-docs/RFCs/v7.1/shader-module-injection-rfc.md

```diff
@@ -0,0 +1,71 @@
+# RFC: Shader Module Code Injection
+
+* **Author**: Tarek Sherif
+* **Date**: May, 2019
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC specifies a mechanism to allow shader modules to inject custom code into the source they are augmenting based on a defined set of directives.
+
+
+## Background
+
+Shader modules currently only add variables and define functions in shaders. Without the ability to inject code in the `main` function, shaders that use them are forced modify their code to use the functionality that a shader module provides.
+
+There is a separate code-injection mechanism in `BaseModel` that allows for code injection by the application. It implements most of the functionality required for module code injection, but limits injections to variable declaration segments, or the beginning or end of the `main` function. This leads to a relatively brittle system, where the application must write arbitrary GLSL code against shader code that might change at some point, and that it cannot affect any intermediate calculations.
+
+
+## Customers
+
+deck.gl's effects system is currently unscalable due to the fact that effects are implemented as shader modules and enabling effects currently requires that shaders in all layers be updated to call module functions.
+
+
+## Overview
+
+The foundation of the proposed system is a set of directives that can be used to mark up shader code and indicate when certain data is ready to be processed, e.g.:
+- `##FRAGMENT_COLOR(color)`
+- `##PICK_COLOR(color)`
+- `##VIEW_MATRIX(matrix)`
+- `##SURFACE_LIGHT(surfaceColor, normal, eyePosition)`
```

Comment:
the arguments in the directives, which is return value and which are function parameters? how do we apply it to the real function?

Reply:
It's better not to think of them as arguments per-se (maybe the API shouldn't be function like, if that causes confusion). It's just means those variables are involved in the text-replacement rule that the injection defines. 

Or on a conceptual level, I'm thinking of directives as being tags that indicate that some data is ready to be used/modified by shader module code.